src/Hedgehog/Extras/Test/New/Monad/PropertyIO.hs

```diff
@@ -0,0 +1,114 @@
+{-# LANGUAGE BlockArguments #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE ImportQualifiedPost #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Hedgehog.Extras.Test.New.Monad.PropertyIO
+  ( PropertyIO(..)
+  , PropertyEnv(..)
+  , propertyIOToPropertyT
+  , forAll
+  ) where
+
+import Control.Concurrent qualified as IO
+import Control.Concurrent.STM
+import Control.Concurrent.STM qualified as STM
+import Control.Exception as IO
+import Control.Exception.Lifted qualified as CEL
+import Control.Monad.Catch (MonadCatch, MonadMask, MonadThrow)
+import Control.Monad.Error.Class (MonadError)
+import Control.Monad.Fix (MonadFix)
+import Control.Monad.Primitive (PrimMonad)
+import Control.Monad.Reader (MonadReader(..))
+import Control.Monad.Trans.Reader (ReaderT(..))
+import Control.Monad.Trans.Resource (ResourceT, MonadResource(..), runResourceT)
+import Data.Either
+import Data.Generics.Product.Any
+import Data.Maybe
+import Data.Monoid
+import HaskellWorks.Control.Monad
+import HaskellWorks.Prelude
+import Hedgehog hiding (forAll)
+import Hedgehog qualified as H
+import Hedgehog.Extras.Test.MonadAssertion
+import Hedgehog.Internal.Property qualified as H
+import Lens.Micro
+import Test.Tasty.Discover
+import Test.Tasty.Hedgehog (testProperty)
+
+newtype PropertyEnv = PropertyEnv { unPropertyEnv :: TMVar (PropertyT IO ()) }
+
+newtype PropertyIO a = PropertyIO { runPropertIO :: ExceptT H.Failure (ResourceT (ReaderT PropertyEnv IO)) a }
+  deriving newtype (Applicative)
+  deriving newtype (Functor)
+  deriving newtype (Monad)
+  deriving newtype (MonadCatch)
+  deriving newtype (MonadError H.Failure)
+  deriving newtype (MonadFail)
+  deriving newtype (MonadFix)
+  deriving newtype (MonadIO)
+  deriving newtype (MonadMask)
+  deriving newtype (MonadReader PropertyEnv)
+  deriving newtype (MonadResource)
+  deriving newtype (MonadThrow)
+  deriving newtype (PrimMonad)
+
+instance MonadTest PropertyIO where
+  liftTest f = do
+    PropertyEnv tmvAction <- ask @PropertyEnv
+    mvA <- liftIO STM.newEmptyTMVarIO
+    liftIO $ STM.atomically $ putTMVar tmvAction $
+      tryExceptAssertion (liftTest f) >>= liftIO . STM.atomically . STM.putTMVar mvA
+    testResult <- liftIO $ STM.atomically $ STM.takeTMVar mvA
```

Comment:
why this TMVar is necessary here? It's not used outside of the `liftTest`.

Reply:
I've added the explanation in comments.